File: MPChartLib/src/test/java/com/github/mikephil/charting/test/DataSetTest.java
package com.github.mikephil.charting.test;

import com.github.mikephil.charting.data.DataSet;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.ScatterDataSet;

import org.junit.Test;

import java.util.ArrayList;
import java.util.List;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertTrue;

/**
 * Created by philipp on 31/05/16.
 */
public class DataSetTest {

    @Test
    public void testCalcMinMax() {

        List<Entry> entries = new ArrayList<Entry>();
        entries.add(new Entry(10, 10));
        entries.add(new Entry(15, 2));
        entries.add(new Entry(21, 5));

        ScatterDataSet set = new ScatterDataSet(entries, "");

        assertEquals(10f, set.getXMin(), 0.01f);
        assertEquals(21f, set.getXMax(), 0.01f);

        assertEquals(2f, set.getYMin(), 0.01f);
        assertEquals(10f, set.getYMax(), 0.01f);

        assertEquals(3, set.getEntryCount());

        set.addEntry(new Entry(25, 1));

        assertEquals(10f, set.getXMin(), 0.01f);
        assertEquals(25f, set.getXMax(), 0.01f);

        assertEquals(1f, set.getYMin(), 0.01f);
        assertEquals(10f, set.getYMax(), 0.01f);

        assertEquals(4, set.getEntryCount());

        set.removeEntry(3);

        assertEquals(10f, set.getXMin(), 0.01f);
        assertEquals(21, set.getXMax(), 0.01f);

        assertEquals(2f, set.getYMin(), 0.01f);
        assertEquals(10f, set.getYMax(), 0.01f);
    }

    @Test
    public void testAddRemoveEntry() {

        List<Entry> entries = new ArrayList<Entry>();
        entries.add(new Entry(10, 10));
        entries.add(new Entry(15, 2));
        entries.add(new Entry(21, 5));

        ScatterDataSet set = new ScatterDataSet(entries, "");

        assertEquals(3, set.getEntryCount());

        set.addEntryOrdered(new Entry(5, 1));

        assertEquals(4, set.getEntryCount());

        assertEquals(5, set.getXMin(), 0.01f);
        assertEquals(21, set.getXMax(), 0.01f);

        assertEquals(1f, set.getYMin(), 0.01f);
        assertEquals(10f, set.getYMax(), 0.01f);

        assertEquals(5, set.getEntryForIndex(0).getX(), 0.01f);
        assertEquals(1, set.getEntryForIndex(0).getY(), 0.01f);

        set.addEntryOrdered(new Entry(20, 50));

        assertEquals(5, set.getEntryCount());

        assertEquals(20, set.getEntryForIndex(3).getX(), 0.01f);
        assertEquals(50, set.getEntryForIndex(3).getY(), 0.01f);

        assertTrue(set.removeEntry(3));

        assertEquals(4, set.getEntryCount());

        assertEquals(21, set.getEntryForIndex(3).getX(), 0.01f);
        assertEquals(5, set.getEntryForIndex(3).getY(), 0.01f);

        assertEquals(5, set.getEntryForIndex(0).getX(), 0.01f);
        assertEquals(1, set.getEntryForIndex(0).getY(), 0.01f);

        assertTrue(set.removeFirst());

        assertEquals(3, set.getEntryCount());

        assertEquals(10, set.getEntryForIndex(0).getX(), 0.01f);
        assertEquals(10, set.getEntryForIndex(0).getY(), 0.01f);

        set.addEntryOrdered(new Entry(15, 3));

        assertEquals(4, set.getEntryCount());

        assertEquals(15, set.getEntryForIndex(1).getX(), 0.01f);
        assertEquals(3, set.getEntryForIndex(1).getY(), 0.01f);

        assertEquals(21, set.getEntryForIndex(3).getX(), 0.01f);
        assertEquals(5, set.getEntryForIndex(3).getY(), 0.01f);

        assertTrue(set.removeLast());

        assertEquals(3, set.getEntryCount());

        assertEquals(15, set.getEntryForIndex(2).getX(), 0.01f);
        assertEquals(2, set.getEntryForIndex(2).getY(), 0.01f);

        assertTrue(set.removeLast());

        assertEquals(2, set.getEntryCount());

        assertTrue(set.removeLast());

        assertEquals(1, set.getEntryCount());

        assertEquals(10, set.getEntryForIndex(0).getX(), 0.01f);
        assertEquals(10, set.getEntryForIndex(0).getY(), 0.01f);

        assertTrue(set.removeLast());

        assertEquals(0, set.getEntryCount());

        assertFalse(set.removeLast());
        assertFalse(set.removeFirst());
    }

    @Test
    public void testGetEntryForXValue() {

        List<Entry> entries = new ArrayList<Entry>();
        entries.add(new Entry(10, 10));
        entries.add(new Entry(15, 5));
        entries.add(new Entry(21, 5));

        ScatterDataSet set = new ScatterDataSet(entries, "");

        Entry closest = set.getEntryForXValue(17, DataSet.Rounding.CLOSEST);
        assertEquals(15, closest.getX(), 0.01f);
        assertEquals(5, closest.getY(), 0.01f);

        closest = set.getEntryForXValue(17, DataSet.Rounding.DOWN);
        assertEquals(15, closest.getX(), 0.01f);
        assertEquals(5, closest.getY(), 0.01f);

        closest = set.getEntryForXValue(15, DataSet.Rounding.DOWN);
        assertEquals(15, closest.getX(), 0.01f);
        assertEquals(5, closest.getY(), 0.01f);

        closest = set.getEntryForXValue(14, DataSet.Rounding.DOWN);
        assertEquals(10, closest.getX(), 0.01f);
        assertEquals(10, closest.getY(), 0.01f);

        closest = set.getEntryForXValue(17, DataSet.Rounding.UP);
        assertEquals(21, closest.getX(), 0.01f);
        assertEquals(5, closest.getY(), 0.01f);

        closest = set.getEntryForXValue(21, DataSet.Rounding.UP);
        assertEquals(21, closest.getX(), 0.01f);
        assertEquals(5, closest.getY(), 0.01f);

        closest = set.getEntryForXValue(21, DataSet.Rounding.CLOSEST);
        assertEquals(21, closest.getX(), 0.01f);
        assertEquals(5, closest.getY(), 0.01f);
    }

    @Test
    public void testGetEntryForXValueWithDuplicates() {

        // sorted list of values (by x position)
        List<Entry> values = new ArrayList<Entry>();
        values.add(new Entry(0, 10));
        values.add(new Entry(1, 20));
        values.add(new Entry(2, 30));
        values.add(new Entry(3, 40));
        values.add(new Entry(3, 50)); // duplicate
        values.add(new Entry(4, 60));
        values.add(new Entry(4, 70)); // duplicate
        values.add(new Entry(5, 80));
        values.add(new Entry(6, 90));
        values.add(new Entry(7, 100));
        values.add(new Entry(8, 110));
        values.add(new Entry(8, 120)); // duplicate

        ScatterDataSet set = new ScatterDataSet(values, "");

        Entry closest = set.getEntryForXValue(0, DataSet.Rounding.CLOSEST);
        assertEquals(0, closest.getX(), 0.01f);
        assertEquals(10, closest.getY(), 0.01f);

        closest = set.getEntryForXValue(5, DataSet.Rounding.CLOSEST);
        assertEquals(5, closest.getX(), 0.01f);
        assertEquals(80, closest.getY(), 0.01f);

        closest = set.getEntryForXValue(5.4f, DataSet.Rounding.CLOSEST);
        assertEquals(5, closest.getX(), 0.01f);
        assertEquals(80, closest.getY(), 0.01f);

        closest = set.getEntryForXValue(4.6f, DataSet.Rounding.CLOSEST);
        assertEquals(5, closest.getX(), 0.01f);
        assertEquals(80, closest.getY(), 0.01f);

        closest = set.getEntryForXValue(7, DataSet.Rounding.CLOSEST);
        assertEquals(7, closest.getX(), 0.01f);
        assertEquals(100, closest.getY(), 0.01f);

        closest = set.getEntryForXValue(4f, DataSet.Rounding.CLOSEST);
        assertEquals(4, closest.getX(), 0.01f);
        assertEquals(60, closest.getY(), 0.01f);

        List<Entry> entries = set.getEntriesForXValue(4f);
        assertEquals(2, entries.size());
        assertEquals(60, entries.get(0).getY(), 0.01f);
        assertEquals(70, entries.get(1).getY(), 0.01f);

        entries = set.getEntriesForXValue(3.5f);
        assertEquals(0, entries.size());

        entries = set.getEntriesForXValue(2f);
        assertEquals(1, entries.size());
        assertEquals(30, entries.get(0).getY(), 0.01f);
    }
}

File: MPChartLib/src/test/java/com/github/mikephil/charting/test/DataSetTest.kt
package com.github.mikephil.charting.test

import com.github.mikephil.charting.data.DataSet
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.data.ScatterDataSet
import org.junit.Assert
import org.junit.Test

class DataSetTest {
    @Test
    fun testCalcMinMax() {
        val entries: MutableList<Entry> = ArrayList()
        entries.add(Entry(10f, 10f))
        entries.add(Entry(15f, 2f))
        entries.add(Entry(21f, 5f))

        val set = ScatterDataSet(entries, "")

        Assert.assertEquals(10f, set.xMin, 0.01f)
        Assert.assertEquals(21f, set.xMax, 0.01f)

        Assert.assertEquals(2f, set.yMin, 0.01f)
        Assert.assertEquals(10f, set.yMax, 0.01f)

        Assert.assertEquals(3, set.entryCount)

        set.addEntry(Entry(25f, 1f))

        Assert.assertEquals(10f, set.xMin, 0.01f)
        Assert.assertEquals(25f, set.xMax, 0.01f)

        Assert.assertEquals(1f, set.yMin, 0.01f)
        Assert.assertEquals(10f, set.yMax, 0.01f)

        Assert.assertEquals(4, set.entryCount)

        set.removeEntry(3)

        Assert.assertEquals(10f, set.xMin, 0.01f)
        Assert.assertEquals(21f, set.xMax, 0.01f)

        Assert.assertEquals(2f, set.yMin, 0.01f)
        Assert.assertEquals(10f, set.yMax, 0.01f)
    }

    @Test
    fun testAddRemoveEntry() {
        val entries: MutableList<Entry> = ArrayList()
        entries.add(Entry(10f, 10f))
        entries.add(Entry(15f, 2f))
        entries.add(Entry(21f, 5f))

        val set = ScatterDataSet(entries, "")

        Assert.assertEquals(3, set.entryCount)

        set.addEntryOrdered(Entry(5f, 1f))

        Assert.assertEquals(4, set.entryCount)

        Assert.assertEquals(5f, set.xMin, 0.01f)
        Assert.assertEquals(21f, set.xMax, 0.01f)

        Assert.assertEquals(1f, set.yMin, 0.01f)
        Assert.assertEquals(10f, set.yMax, 0.01f)

        Assert.assertEquals(5f, set.getEntryForIndex(0).x, 0.01f)
        Assert.assertEquals(1f, set.getEntryForIndex(0).y, 0.01f)

        set.addEntryOrdered(Entry(20f, 50f))

        Assert.assertEquals(5, set.entryCount)

        Assert.assertEquals(20f, set.getEntryForIndex(3).x, 0.01f)
        Assert.assertEquals(50f, set.getEntryForIndex(3).y, 0.01f)

        Assert.assertTrue(set.removeEntry(3))

        Assert.assertEquals(4, set.entryCount)

        Assert.assertEquals(21f, set.getEntryForIndex(3).x, 0.01f)
        Assert.assertEquals(5f, set.getEntryForIndex(3).y, 0.01f)

        Assert.assertEquals(5f, set.getEntryForIndex(0).x, 0.01f)
        Assert.assertEquals(1f, set.getEntryForIndex(0).y, 0.01f)

        Assert.assertTrue(set.removeFirst())

        Assert.assertEquals(3, set.entryCount)

        Assert.assertEquals(10f, set.getEntryForIndex(0).x, 0.01f)
        Assert.assertEquals(10f, set.getEntryForIndex(0).y, 0.01f)

        set.addEntryOrdered(Entry(15f, 3f))

        Assert.assertEquals(4, set.entryCount)

        Assert.assertEquals(15f, set.getEntryForIndex(1).x, 0.01f)
        Assert.assertEquals(3f, set.getEntryForIndex(1).y, 0.01f)

        Assert.assertEquals(21f, set.getEntryForIndex(3).x, 0.01f)
        Assert.assertEquals(5f, set.getEntryForIndex(3).y, 0.01f)

        Assert.assertTrue(set.removeLast())

        Assert.assertEquals(3, set.entryCount)

        Assert.assertEquals(15f, set.getEntryForIndex(2).x, 0.01f)
        Assert.assertEquals(2f, set.getEntryForIndex(2).y, 0.01f)

        Assert.assertTrue(set.removeLast())

        Assert.assertEquals(2, set.entryCount)

        Assert.assertTrue(set.removeLast())

        Assert.assertEquals(1, set.entryCount)

        Assert.assertEquals(10f, set.getEntryForIndex(0).x, 0.01f)
        Assert.assertEquals(10f, set.getEntryForIndex(0).y, 0.01f)

        Assert.assertTrue(set.removeLast())

        Assert.assertEquals(0, set.entryCount)

        Assert.assertFalse(set.removeLast())
        Assert.assertFalse(set.removeFirst())
    }

    @Test
    fun testGetEntryForXValue() {
        val entries: MutableList<Entry> = ArrayList()
        entries.add(Entry(10f, 10f))
        entries.add(Entry(15f, 5f))
        entries.add(Entry(21f, 5f))

        val set = ScatterDataSet(entries, "")

        var closest = set.getEntryForXValue(17f, Float.NaN, DataSet.Rounding.CLOSEST)
        Assert.assertEquals(15f, closest.x, 0.01f)
        Assert.assertEquals(5f, closest.y, 0.01f)

        closest = set.getEntryForXValue(17f, Float.NaN, DataSet.Rounding.DOWN)
        Assert.assertEquals(15f, closest.x, 0.01f)
        Assert.assertEquals(5f, closest.y, 0.01f)

        closest = set.getEntryForXValue(15f, Float.NaN, DataSet.Rounding.DOWN)
        Assert.assertEquals(15f, closest.x, 0.01f)
        Assert.assertEquals(5f, closest.y, 0.01f)

        closest = set.getEntryForXValue(14f, Float.NaN, DataSet.Rounding.DOWN)
        Assert.assertEquals(10f, closest.x, 0.01f)
        Assert.assertEquals(10f, closest.y, 0.01f)

        closest = set.getEntryForXValue(17f, Float.NaN, DataSet.Rounding.UP)
        Assert.assertEquals(21f, closest.x, 0.01f)
        Assert.assertEquals(5f, closest.y, 0.01f)

        closest = set.getEntryForXValue(21f, Float.NaN, DataSet.Rounding.UP)
        Assert.assertEquals(21f, closest.x, 0.01f)
        Assert.assertEquals(5f, closest.y, 0.01f)

        closest = set.getEntryForXValue(21f, Float.NaN, DataSet.Rounding.CLOSEST)
        Assert.assertEquals(21f, closest.x, 0.01f)
        Assert.assertEquals(5f, closest.y, 0.01f)
    }

    @Test
    fun testGetEntryForXValueWithDuplicates() {
        // sorted list of values (by x position)

        val values: MutableList<Entry> = ArrayList()
        values.add(Entry(0f, 10f))
        values.add(Entry(1f, 20f))
        values.add(Entry(2f, 30f))
        values.add(Entry(3f, 40f))
        values.add(Entry(3f, 50f)) // duplicate
        values.add(Entry(4f, 60f))
        values.add(Entry(4f, 70f)) // duplicate
        values.add(Entry(5f, 80f))
        values.add(Entry(6f, 90f))
        values.add(Entry(7f, 100f))
        values.add(Entry(8f, 110f))
        values.add(Entry(8f, 120f)) // duplicate

        val set = ScatterDataSet(values, "")

        var closest = set.getEntryForXValue(0f, Float.NaN, DataSet.Rounding.CLOSEST)
        Assert.assertEquals(0f, closest.x, 0.01f)
        Assert.assertEquals(10f, closest.y, 0.01f)

        closest = set.getEntryForXValue(5f, Float.NaN, DataSet.Rounding.CLOSEST)
        Assert.assertEquals(5f, closest.x, 0.01f)
        Assert.assertEquals(80f, closest.y, 0.01f)

        closest = set.getEntryForXValue(5.4f, Float.NaN, DataSet.Rounding.CLOSEST)
        Assert.assertEquals(5f, closest.x, 0.01f)
        Assert.assertEquals(80f, closest.y, 0.01f)

        closest = set.getEntryForXValue(4.6f, Float.NaN, DataSet.Rounding.CLOSEST)
        Assert.assertEquals(5f, closest.x, 0.01f)
        Assert.assertEquals(80f, closest.y, 0.01f)

        closest = set.getEntryForXValue(7f, Float.NaN, DataSet.Rounding.CLOSEST)
        Assert.assertEquals(7f, closest.x, 0.01f)
        Assert.assertEquals(100f, closest.y, 0.01f)

        closest = set.getEntryForXValue(4f, Float.NaN, DataSet.Rounding.CLOSEST)
        Assert.assertEquals(4f, closest.x, 0.01f)
        Assert.assertEquals(60f, closest.y, 0.01f)

        var entries = set.getEntriesForXValue(4f)
        Assert.assertEquals(2, entries.size)
        Assert.assertEquals(60f, entries[0].y, 0.01f)
        Assert.assertEquals(70f, entries[1].y, 0.01f)

        entries = set.getEntriesForXValue(3.5f)
        Assert.assertEquals(0, entries.size)

        entries = set.getEntriesForXValue(2f)
        Assert.assertEquals(1, entries.size)
        Assert.assertEquals(30f, entries[0].y, 0.01f)
    }
}
